crates/cli/src/commands/profile.rs

```diff
@@ -0,0 +1,187 @@
+
+        let mut last_known_time = std::time::Instant::now();
+        for sample in samples {
+            let mut frames = Vec::new();
+            for frame in sample.stack {
+                frames.push(Frame::Label(thread_builder.intern_string(&frame)))
+            }
+
+            thread_builder.add_sample(
+                last_known_time,
+                frames.into_iter(),
+                // this is actually in instructions but
+                std::time::Duration::from_micros(1),
+            );
+
+            last_known_time += std::time::Duration::from_micros(1);
+        }
+
+        profile_builder.add_thread(thread_builder);
+
+        let mut file = std::fs::File::create(&self.output)?;
+        serde_json::to_writer(&mut file, &profile_builder.to_serializable())?;
+
+        Ok(())
+    }
+}
+
+fn build_goblin_lookups(
+    start_lookup: &mut HashMap<u64, String>,
+    end_lookup: &mut HashMap<u64, String>,
+    func_range_lookup: &mut HashMap<String, (u64, u64)>,
+    elf_name: &str,
+) -> std::io::Result<()> {
+    let buffer = std::fs::read(elf_name).unwrap();
+    let elf = Elf::parse(&buffer).unwrap();
+
+    for sym in &elf.syms {
+        if sym.st_type() == STT_FUNC {
+            let name = elf.strtab.get_at(sym.st_name).unwrap_or("");
+            let demangled_name = demangle(name);
+            let size = sym.st_size;
+            let start_address = sym.st_value;
+            let end_address = start_address + size - 4;
+            start_lookup.insert(start_address, demangled_name.to_string());
+            end_lookup.insert(end_address, demangled_name.to_string());
+            func_range_lookup.insert(demangled_name.to_string(), (start_address, end_address));
+        }
+    }
+    Ok(())
+}
+
+pub fn collect_samples(
+    elf_path: impl AsRef<Path>,
+    trace_path: impl AsRef<Path>,
+) -> Result<Vec<Sample>> {
+    let elf_path = elf_path.as_ref();
+    let trace_path = trace_path.as_ref();
+
+    let mut start_lookup = HashMap::new();
+    let mut end_lookup = HashMap::new();
+    let mut func_range_lookup = HashMap::new();
+    build_goblin_lookups(
+        &mut start_lookup,
+        &mut end_lookup,
+        &mut func_range_lookup,
+        &elf_path.to_string_lossy(),
+    )
+    .unwrap();
+
+    let mut function_ranges: Vec<(u64, u64, String)> =
+        func_range_lookup.iter().map(|(f, &(start, end))| (start, end, f.clone())).collect();
+
+    function_ranges.sort_by_key(|&(start, _, _)| start);
+
+    let file = std::fs::File::open(trace_path).unwrap();
+    let file_size = file.metadata().unwrap().len();
+    let mut buf = std::io::BufReader::new(file);
+    let mut function_stack: Vec<Rc<str>> = Vec::new();
+    let total_lines = file_size / 4;
+    let mut current_function_range: (u64, u64) = (0, 0);
+
+    // unsafe on 32 bit systems but ...
+    let mut samples = Vec::with_capacity(total_lines as usize);
+    for _ in 0..total_lines {
```

Comment:
I dont think we actually need to record a sample per instruction

I think we could instead store the instruction count between *changes* in the stack?

Reply:
Im still unclear on exactly how the flamegraph interprets time 

what is cpu usage delta?
is the flamegraph just created by looking at the total number of samples over how many samples is it this call stack? Not sure how they would differentiate calls in a loop